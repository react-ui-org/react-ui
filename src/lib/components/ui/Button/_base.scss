@use '../../../styles/tools/accessibility';
@use '../../../styles/tools/breakpoint';
@use 'settings';
@use 'theme';
@use 'tools';

.root {
  @include tools.button();
}

.label {
  display: block;
}

.beforeLabel,
.afterLabel,
.startCorner,
.endCorner,
.loadingIcon {
  display: flex;
  align-items: baseline;
  justify-content: center;
}

.beforeLabel {
  margin-right: settings.$icon-spacing;
}

.afterLabel,
.loadingIcon {
  margin-left: settings.$icon-spacing;
}

.startCorner,
.endCorner {
  position: absolute;
  top: -0.35rem;
}

.startCorner {
  left: 0;
  margin-left: -0.35rem;
}

.endCorner {
  right: 0;
  margin-right: -0.35rem;
}

.rootSizeSmall {
  @include tools.button-size(small);
}

.rootSizeMedium {
  @include tools.button-size(medium);
}

.rootSizeLarge {
  @include tools.button-size(large);
}

.rootBlock {
  width: 100%;
}

.rootBlock.withLabelHiddenMobile {
  @include breakpoint.up(settings.$breakpoint) {
    width: 100%;
  }
}

.rootGrouped {
  z-index: map-get(settings.$group-z-indexes, button);

  &:not(:first-child) {
    margin-left: var(--local-gap);
    border-top-left-radius: 0;
    border-bottom-left-radius: 0;
  }

  &:not(:last-child) {
    border-top-right-radius: 0;
    border-bottom-right-radius: 0;
  }

  &:focus,
  &:not(:disabled):hover {
    z-index: map-get(settings.$group-z-indexes, button-hover);
  }
}

.rootGrouped .startCorner,
.rootGrouped .endCorner {
  z-index: map-get(settings.$group-z-indexes, button-overflowing-elements);
}

.withLabelHidden .label,
.withLabelHiddenMobile .label {
  @include accessibility.hide-text();
}

.withLabelHiddenMobile .label {
  @include breakpoint.up(settings.$breakpoint) {
    position: relative;
    width: auto;
    height: auto;
  }
}

.withLabelHidden .beforeLabel,
.withLabelHidden .afterLabel,
.withLabelHiddenMobile .beforeLabel,
.withLabelHiddenMobile .afterLabel {
  margin-right: 0;
  margin-left: 0;
}

.withLabelHiddenMobile .beforeLabel {
  @include breakpoint.up(settings.$breakpoint) {
    margin-right: settings.$icon-spacing;
  }
}

.withLabelHiddenMobile .afterLabel {
  @include breakpoint.up(settings.$breakpoint) {
    margin-left: settings.$icon-spacing;
  }
}

.withLabelHidden.isRootLoading .loadingIcon,
.withLabelHiddenMobile.isRootLoading .loadingIcon {
  margin-left: 0;
}

.withLabelHidden.isRootLoading .beforeLabel,
.withLabelHidden.isRootLoading .afterLabel,
.withLabelHiddenMobile.isRootLoading .beforeLabel,
.withLabelHiddenMobile.isRootLoading .afterLabel {
  display: none;
}

.withLabelHiddenMobile.isRootLoading .beforeLabel,
.withLabelHiddenMobile.isRootLoading .afterLabel {
  @include breakpoint.up(settings.$breakpoint) {
    display: flex;
  }
}

.withLabelHiddenMobile.isRootLoading .loadingIcon {
  @include breakpoint.up(settings.$breakpoint) {
    margin-left: settings.$icon-spacing;
  }
}
