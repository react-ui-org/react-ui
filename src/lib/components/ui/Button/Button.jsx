import PropTypes from 'prop-types';
import React from 'react';
import transferProps from '../../../utils/transferProps';
import withForwardedRef from '../withForwardedRef';
import styles from './Button.scss';

export const Button = (props) => {
  const {
    afterLabel,
    beforeLabel,
    block,
    clickHandler,
    disabled,
    endCorner,
    forwardedRef,
    grouped,
    id,
    label,
    labelVisibility,
    loadingIcon,
    priority,
    size,
    startCorner,
    type,
    variant,
  } = props;

  let priorityClass = '';
  let sizeClass = '';
  let variantClass = '';
  let blockClass = '';
  let groupedClass = '';
  let labelVisibilityClass = '';
  let loadingClass = '';

  if (priority === 'default') {
    priorityClass = styles.priorityDefault;
  } else if (priority === 'outline') {
    priorityClass = styles.priorityOutline;
  } else if (priority === 'flat') {
    priorityClass = styles.priorityFlat;
  } else if (priority === 'link') {
    priorityClass = styles.priorityLink;
  }

  if (priority !== 'link') {
    if (size === 'small') {
      sizeClass = styles.sizeSmall;
    } else if (size === 'medium') {
      sizeClass = styles.sizeMedium;
    } else if (size === 'large') {
      sizeClass = styles.sizeLarge;
    }

    if (variant === 'primary') {
      variantClass = styles.variantPrimary;
    } else if (variant === 'secondary') {
      variantClass = styles.variantSecondary;
    } else if (variant === 'success') {
      variantClass = styles.variantSuccess;
    } else if (variant === 'warning') {
      variantClass = styles.variantWarning;
    } else if (variant === 'danger') {
      variantClass = styles.variantDanger;
    } else if (variant === 'dark') {
      variantClass = styles.variantDark;
    }

    if (block) {
      blockClass = styles.isRootBlock;
    }

    if (grouped) {
      groupedClass = styles.isRootGrouped;
    }

    if (labelVisibility === 'desktop') {
      labelVisibilityClass = styles.withLabelHiddenMobile;
    } else if (labelVisibility === 'none') {
      labelVisibilityClass = styles.withLabelHidden;
    }

    if (loadingIcon) {
      loadingClass = styles.isRootLoading;
    }
  }

  const propsToTransfer = transferProps(
    props,
    ['afterLabel', 'beforeLabel', 'block', 'clickHandler', 'disabled', 'endCorner', 'grouped', 'id',
      'label', 'labelVisibility', 'loadingIcon', 'priority', 'size', 'startCorner', 'type', 'variant'],
  );

  return (
    <button
      {...propsToTransfer}
      className={(`
        ${styles.root}
        ${priorityClass}
        ${sizeClass}
        ${variantClass}
        ${blockClass}
        ${groupedClass}
        ${labelVisibilityClass}
        ${loadingClass}
      `).trim()}
      disabled={disabled || !!loadingIcon}
      id={id}
      onClick={clickHandler}
      ref={forwardedRef}
      type={type}
    >
      {priority !== 'link' && startCorner && (
        <span className={styles.startCorner}>
          {startCorner}
        </span>
      )}
      {beforeLabel && (
        <span className={styles.beforeLabel}>
          {beforeLabel}
        </span>
      )}
      <span
        className={styles.label}
        {...(id && { id: `${id}__labelText` })}
      >
        {label}
      </span>
      {afterLabel && (
        <span className={styles.afterLabel}>
          {afterLabel}
        </span>
      )}
      {priority !== 'link' && endCorner && (
        <span className={styles.endCorner}>
          {endCorner}
        </span>
      )}
      {priority !== 'link' && loadingIcon && (
        <span className={styles.loadingIcon}>
          {loadingIcon}
        </span>
      )}
    </button>
  );
};

Button.defaultProps = {
  afterLabel: null,
  beforeLabel: null,
  block: false,
  clickHandler: null,
  disabled: false,
  endCorner: null,
  forwardedRef: undefined,
  grouped: false,
  id: undefined,
  labelVisibility: 'all',
  loadingIcon: null,
  priority: 'default',
  size: 'medium',
  startCorner: null,
  type: 'button',
  variant: 'primary',
};

Button.propTypes = {
  afterLabel: PropTypes.node,
  beforeLabel: PropTypes.node,
  block: PropTypes.bool,
  clickHandler: PropTypes.func,
  disabled: PropTypes.bool,
  endCorner: PropTypes.node,
  forwardedRef: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({ current: PropTypes.any }),
  ]),
  grouped: PropTypes.bool,
  id: PropTypes.string,
  label: PropTypes.string.isRequired,
  labelVisibility: PropTypes.oneOf(['all', 'desktop', 'none']),
  loadingIcon: PropTypes.node,
  priority: PropTypes.oneOf(['default', 'outline', 'flat', 'link']),
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  startCorner: PropTypes.node,
  type: PropTypes.oneOf(['button', 'submit']),
  variant: PropTypes.oneOf(['primary', 'secondary', 'success', 'warning', 'danger', 'dark']),
};

export default withForwardedRef(Button);
