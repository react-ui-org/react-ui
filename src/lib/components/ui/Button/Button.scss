@import '../../../styles/settings/colors';
@import '../../../styles/tools/accessibility';
@import '../../../styles/tools/breakpoints';
@import '../../../styles/tools/forms/relationships';
@import 'mixins';
@import 'variables';

.root {
  @include button();
  @include form-field-horizontal-neighbor();
}

.label {
  display: block;
}

.beforeLabel,
.afterLabel,
.startCorner,
.endCorner,
.loadingIcon {
  display: flex;
  align-items: baseline;
  justify-content: center;
}

.beforeLabel {
  margin-right: $button-icon-offset;
}

.afterLabel,
.loadingIcon {
  margin-left: $button-icon-offset;
}

.startCorner,
.endCorner {
  position: absolute;
  top: -0.35rem;
}

.startCorner {
  left: 0;
  margin-left: -0.35rem;
}

.endCorner {
  right: 0;
  margin-right: -0.35rem;
}

.sizeSmall {
  @include button-size(small);
}

.sizeMedium {
  @include button-size(medium);
}

.sizeLarge {
  @include button-size(large);
}

.isRootBlock {
  @include form-field-vertical-neighbor();

  width: 100%;
}

.isRootBlock.withLabelHiddenMobile {
  @include breakpoint-up($button-breakpoint) {
    width: 100%;
  }
}

.isRootGrouped {
  &:not(:first-child) {
    margin-left: calc(-1 * #{$button-border-width});
    border-top-left-radius: 0;
    border-bottom-left-radius: 0;
  }

  &:not(:last-child) {
    margin-right: 0;
    border-top-right-radius: 0;
    border-bottom-right-radius: 0;
  }
}

.withLabelHidden .label,
.withLabelHiddenMobile .label {
  @include accessibility-hide-text();
}

.withLabelHiddenMobile .label {
  @include breakpoint-up($button-breakpoint) {
    position: relative;
    width: auto;
    height: auto;
  }
}

.withLabelHidden .beforeLabel,
.withLabelHidden .afterLabel,
.withLabelHiddenMobile .beforeLabel,
.withLabelHiddenMobile .afterLabel {
  margin-right: 0;
  margin-left: 0;
}

.withLabelHiddenMobile .beforeLabel {
  @include breakpoint-up($button-breakpoint) {
    margin-right: $button-icon-offset;
  }
}

.withLabelHiddenMobile .afterLabel {
  @include breakpoint-up($button-breakpoint) {
    margin-left: $button-icon-offset;
  }
}

// Default priority variants
.priorityDefault.variantPrimary {
  @include button-variant(default, primary);
}

.priorityDefault.variantSecondary {
  @include button-variant(default, secondary);
}

.priorityDefault.variantSuccess {
  @include button-variant(default, success);
}

.priorityDefault.variantWarning {
  @include button-variant(default, warning);
}

.priorityDefault.variantDanger {
  @include button-variant(default, danger);
}

.priorityDefault.variantDark {
  @include button-variant(default, dark);
}

// Outline priority variants
.priorityOutline.variantPrimary {
  @include button-variant(outline, primary);
}

.priorityOutline.variantSecondary {
  @include button-variant(outline, secondary);
}

.priorityOutline.variantSuccess {
  @include button-variant(outline, success);
}

.priorityOutline.variantWarning {
  @include button-variant(outline, warning);
}

.priorityOutline.variantDanger {
  @include button-variant(outline, danger);
}

.priorityOutline.variantDark {
  @include button-variant(outline, dark);
}

// Flat priority variants
.priorityFlat.variantPrimary {
  @include button-variant(flat, primary);
}

.priorityFlat.variantSecondary {
  @include button-variant(flat, secondary);
}

.priorityFlat.variantSuccess {
  @include button-variant(flat, success);
}

.priorityFlat.variantWarning {
  @include button-variant(flat, warning);
}

.priorityFlat.variantDanger {
  @include button-variant(flat, danger);
}

.priorityFlat.variantDark {
  @include button-variant(flat, dark);
}

// Grouped buttons
.priorityDefault.isRootGrouped:not(:first-child)::before,
.priorityFlat.isRootGrouped:not(:first-child)::before {
  content: '';
  position: absolute;
  top: calc(-1 * #{$button-border-width});
  bottom: calc(-1 * #{$button-border-width});
  left: calc(-1 * #{$button-border-width});
  border-left: $button-border-width solid $color-white;
}

.priorityFlat.isRootGrouped:not(:first-child)::before {
  border-left-color: currentColor;
}
