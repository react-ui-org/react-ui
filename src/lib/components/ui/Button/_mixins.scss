// 1. Get ready for position of corner elements and group separator.
// 2. Icon buttons should be squares.

@import '../../../styles/tools/explode';
@import '../../../styles/tools/forms/states';
@import '../../../styles/tools/transitions';
@import 'variables';
@import 'theme';

@mixin button() {
  @include transition((opacity, color, border, background, box-shadow));

  position: relative; // 1.
  display: inline-flex;
  align-items: center;
  justify-content: center;
  font-weight: $button-font-weight;
  line-height: $button-line-height;
  font-family: $button-font-family;
  text-align: center;
  text-overflow: ellipsis;
  text-decoration: none;
  text-transform: $button-text-transform;
  white-space: nowrap;
  vertical-align: middle;
  border: $button-border-width solid transparent;
  border-radius: $button-border-radius;
  user-select: none;
  cursor: pointer;

  &:hover,
  &:focus {
    text-decoration: none;
  }

  &:disabled {
    @include form-field-disabled();
  }

  &:-moz-focusring {
    outline: 0;
  }
}

@mixin button-size($size) {
  $properties: map-get($button-sizes, $size);

  height: map-get($properties, height);
  padding: 0 map-get($properties, padding-x);
  font-size: map-get($properties, font-size);

  &.withLabelHidden,
  &.withLabelHiddenMobile {
    width: map-get($properties, height); // 2.
    padding-right: 0;
    padding-left: 0;
  }

  &.withLabelHiddenMobile {
    @include breakpoint-up($button-breakpoint) {
      width: auto;
      padding-right: map-get($properties, padding-x);
      padding-left: map-get($properties, padding-x);
    }
  }
}

@mixin button-variant($priority, $variant) {
  $priority-variants: map-get($button-variants, $priority);
  $variant-states: map-get($priority-variants, $variant);

  @include explode(map-get($variant-states, default));

  &:hover,
  &:focus {
    @include explode(map-get($variant-states, hover));
  }

  &:active {
    @include explode(map-get($variant-states, active));
  }

  &:disabled {
    @include explode(map-get($variant-states, default));
  }
}
