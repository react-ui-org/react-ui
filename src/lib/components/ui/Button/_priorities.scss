@import 'mixins';
@import 'theme';

.priorityDefault.variantPrimary {
  @include button-variant(default, primary);
}

.priorityDefault.variantSecondary {
  @include button-variant(default, secondary);
}

.priorityDefault.variantSuccess {
  @include button-variant(default, success);
}

.priorityDefault.variantWarning {
  @include button-variant(default, warning);
}

.priorityDefault.variantDanger {
  @include button-variant(default, danger);
}

.priorityDefault.variantDark {
  @include button-variant(default, dark);
}

.priorityOutline.variantPrimary {
  @include button-variant(outline, primary);
}

.priorityOutline.variantSecondary {
  @include button-variant(outline, secondary);
}

.priorityOutline.variantSuccess {
  @include button-variant(outline, success);
}

.priorityOutline.variantWarning {
  @include button-variant(outline, warning);
}

.priorityOutline.variantDanger {
  @include button-variant(outline, danger);
}

.priorityOutline.variantDark {
  @include button-variant(outline, dark);
}

.priorityFlat.variantPrimary {
  @include button-variant(flat, primary);
}

.priorityFlat.variantSecondary {
  @include button-variant(flat, secondary);
}

.priorityFlat.variantSuccess {
  @include button-variant(flat, success);
}

.priorityFlat.variantWarning {
  @include button-variant(flat, warning);
}

.priorityFlat.variantDanger {
  @include button-variant(flat, danger);
}

.priorityFlat.variantDark {
  @include button-variant(flat, dark);
}

.priorityLink {
  @include button-link();
}

.priorityDefault.isRootGrouped:not(:first-child)::before,
.priorityFlat.isRootGrouped:not(:first-child)::before {
  content: '';
  position: absolute;
  top: calc(-1 * #{$button-border-width});
  bottom: calc(-1 * #{$button-border-width});
  left: calc(-1 * #{$button-border-width});
  border-left: $button-border-width solid $button-group-separator-color;
}

.priorityFlat.isRootGrouped:not(:first-child)::before {
  border-left-color: currentColor;
}
