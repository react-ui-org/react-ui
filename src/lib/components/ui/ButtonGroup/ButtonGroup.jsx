import PropTypes from 'prop-types';
import React from 'react';
import styles from './ButtonGroup.scss';

export const ButtonGroup = (props) => {
  const {
    block,
    disabled,
    children,
    priority,
    size,
    ...other
  } = props;

  let blockClass = '';
  if (block) {
    blockClass = styles.isRootBlock;
  }

  return (
    <div
      className={`
        ${styles.root}
        ${blockClass}
      `.trim()}
      role="group"
      {...other}
    >
      {React.Children.map(children, (child) => {
        if (!React.isValidElement(child)) {
          return null;
        }

        return React.cloneElement(child, {
          block,
          disabled,
          grouped: true,
          priority,
          size,
        });
      })}
    </div>
  );
};

ButtonGroup.defaultProps = {
  block: false,
  disabled: false,
  priority: 'default',
  size: 'medium',
};

ButtonGroup.propTypes = {
  /**
   * If `true`, the button group will span the full width of its parent.
   */
  block: PropTypes.bool,
  /**
   * Buttons to be grouped.
   */
  children: PropTypes.arrayOf(PropTypes.element).isRequired,
  /**
   * If `true`, all buttons inside the group will be disabled.
   */
  disabled: PropTypes.bool,
  /**
   * Visual priority to highlight or suppress the buttons.
   */
  priority: PropTypes.oneOf(['default', 'outline', 'flat']),
  /**
   * Size of the buttons.
   */
  size: PropTypes.oneOf(['small', 'medium', 'large']),
};

export default ButtonGroup;
