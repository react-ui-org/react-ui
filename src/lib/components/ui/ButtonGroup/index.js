export { default as ButtonGroup } from './ButtonGroup';
export { ButtonGroupContext } from './ButtonGroupContext';
