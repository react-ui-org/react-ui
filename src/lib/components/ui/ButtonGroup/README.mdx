---
name: ButtonGroup
menu: 'Components: UI'
route: /components/ui/button-group
---

# ButtonGroup

ButtonGroup visually groups related buttons together.

import {
  Playground,
  Props,
} from 'docz'
import { Button } from '../Button/Button'
import { ButtonGroup } from './ButtonGroup'

## Basic Usage

To implement the ButtonGroup component, you need to import it first:

```js
import { ButtonGroup } from '@react-ui-org/react-ui';
```

And use it:

<Playground>
  <ButtonGroup>
    <Button label="Week" />
    <Button label="Month" />
    <Button label="Year" />
  </ButtonGroup>
</Playground>

See [API](#api) for all available options.

## General Guidelines

- Use button group to group **related actions** that a user can take. Buttons
  should not be grouped just to save space on the screen.

- Use **short labels or icons** so the buttons can fit small screens.

- For toggling between on/off states, use rather the
  [Toggle](/components/ui/toggle) component.

- For switching between options in a form that needs to be submitted, use rather
  the [SelectField](/components/ui/select-field) or
  [Radio](/components/ui/radio) components.

### Shared Properties

You can set the following properties directly on ButtonGroup to be shared for
all buttons inside the group: size, priority, disabled state, and block width.
These properties are then passed over to individual buttons. At the same time,
they **cannot be overridden** on the buttons' level. While technically possible,
from the design point of view it's undesirable to group buttons of totally
different types or sizes.

## Priorities

There are three **visual priorities** of buttons which exactly copy the
priorities of the [Button](/components/ui/button) component:

1. filled
2. outline
3. flat

### Filled

The default, high-emphasis priority should be used for primary actions of your
app.

<Playground>
  <ButtonGroup>
    <Button label="Week" />
    <Button label="Month" />
    <Button label="Year" />
  </ButtonGroup>
</Playground>

### Outline

Medium-emphasis buttons to contain actions that are important but not primary
in your app.

<Playground>
  <ButtonGroup priority="outline">
    <Button label="Week" />
    <Button label="Month" />
    <Button label="Year" />
  </ButtonGroup>
</Playground>

### Flat

Flat buttons are designed for less pronounced actions to help maintain focus on
the content.

<Playground>
  <ButtonGroup priority="flat">
    <Button label="Week" />
    <Button label="Month" />
    <Button label="Year" />
  </ButtonGroup>
</Playground>

## Sizes

All existing button sizes are also available on the button group level: small,
medium, and large.

<Playground>
  <ButtonGroup size="small">
    <Button label="Week" />
    <Button label="Month" />
    <Button label="Year" />
  </ButtonGroup>
  <ButtonGroup>
    <Button label="Week" />
    <Button label="Month" />
    <Button label="Year" />
  </ButtonGroup>
  <ButtonGroup size="large">
    <Button label="Week" />
    <Button label="Month" />
    <Button label="Year" />
  </ButtonGroup>
</Playground>

Block button groups span the full width of a parent:

<Playground>
  <ButtonGroup block>
    <Button label="Week" />
    <Button label="Month" />
    <Button label="Year" />
  </ButtonGroup>
</Playground>

## States

### Disabled State

Disables all buttons inside the group.

<Playground>
  <ButtonGroup disabled>
    <Button label="Week" />
    <Button label="Month" />
    <Button label="Year" />
  </ButtonGroup>
  <ButtonGroup priority="outline" disabled>
    <Button label="Week" />
    <Button label="Month" />
    <Button label="Year" />
  </ButtonGroup>
  <ButtonGroup priority="flat" disabled>
    <Button label="Week" />
    <Button label="Month" />
    <Button label="Year" />
  </ButtonGroup>
</Playground>

### Active State

To highlight the selected option, the active state can be achieved by setting
a different button variant for the active item.

<Playground>
  <ButtonGroup>
    <Button variant="dark" label="Week" />
    <Button label="Month" />
    <Button label="Year" />
  </ButtonGroup>
  <ButtonGroup priority="outline">
    <Button variant="dark" label="Week" />
    <Button label="Month" />
    <Button label="Year" />
  </ButtonGroup>
  <ButtonGroup priority="flat">
    <Button variant="dark" label="Week" />
    <Button label="Month" />
    <Button label="Year" />
  </ButtonGroup>
</Playground>

## Accessibility

You can improve the accessibility of your ButtonGroup by linking it to a label
and communicating the state of individual options.

<Playground>
  <>
    <span id="period-label">Period:</span>
    <ButtonGroup aria-labelledby="#period-label">
      <Button label="Week" aria-pressed variant="dark" />
      <Button label="Month" aria-pressed={false} />
      <Button label="Year" aria-pressed={false} />
    </ButtonGroup>
  </>
</Playground>

## API

<Props of={ButtonGroup} />

---

Next: [Card â†’](/components/ui/card)
