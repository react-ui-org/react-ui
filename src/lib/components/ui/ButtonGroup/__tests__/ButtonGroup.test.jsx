import React from 'react';
import {
  render,
  within,
} from '@testing-library/react';
import Button from '../../Button';
import { ButtonGroup } from '../ButtonGroup';

const mandatoryProps = {
  children: [<Button label="label" />],
};

describe('rendering', () => {
  it.each([
    [
      { block: true },
      (rootElement) => expect(rootElement).toHaveClass('isRootBlock'),
    ],
    [
      { block: false },
      (rootElement) => expect(rootElement).not.toHaveClass('isRootBlock'),
    ],
    [
      { children: [<Button label="label text" />] },
      (rootElement) => expect(within(rootElement).getByText('label text')),
    ],
    [
      { children: [] },
      (rootElement) => expect(rootElement).toBeInTheDocument(),
    ],
    [
      { disabled: true },
      (rootElement) => expect(within(rootElement).getByRole('button')).toBeDisabled(),
    ],
    [
      { disabled: false },
      (rootElement) => expect(within(rootElement).getByRole('button')).not.toBeDisabled(),
    ],
    [
      { priority: 'filled' },
      (rootElement) => expect(within(rootElement).getByRole('button')).toHaveClass('rootPriorityFilled'),
    ],
    [
      { priority: 'outline' },
      (rootElement) => expect(within(rootElement).getByRole('button')).toHaveClass('rootPriorityOutline'),
    ],
    [
      { priority: 'flat' },
      (rootElement) => expect(within(rootElement).getByRole('button')).toHaveClass('rootPriorityFlat'),
    ],
    [
      { size: 'small' },
      (rootElement) => expect(within(rootElement).getByRole('button')).toHaveClass('rootSizeSmall'),
    ],
    [
      { size: 'medium' },
      (rootElement) => expect(within(rootElement).getByRole('button')).toHaveClass('rootSizeMedium'),
    ],
    [
      { size: 'large' },
      (rootElement) => expect(within(rootElement).getByRole('button')).toHaveClass('rootSizeLarge'),
    ],
  ])('renders with props: "%s"', (testedProps, assert) => {
    const dom = render((
      <ButtonGroup
        {...mandatoryProps}
        {...testedProps}
      />
    ));

    assert(dom.container.firstChild);
  });
});
