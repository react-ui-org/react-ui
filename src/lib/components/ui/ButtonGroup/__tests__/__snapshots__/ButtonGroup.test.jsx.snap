// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly mandatory props only 1`] = `
<div
  className="root"
  role="group"
>
  <ForwardRef(withForwardedRef(Button))
    block={false}
    disabled={false}
    grouped={true}
    key=".0"
    label="One"
    priority="default"
    size="medium"
  />
  <ForwardRef(withForwardedRef(Button))
    block={false}
    disabled={false}
    grouped={true}
    key=".1"
    label="Two"
    priority="default"
    size="medium"
  />
</div>
`;

exports[`rendering renders correctly priority 1`] = `
<div
  className="root"
  role="group"
>
  <ForwardRef(withForwardedRef(Button))
    block={false}
    disabled={false}
    grouped={true}
    key=".0"
    label="One"
    priority="outline"
    size="medium"
  />
  <ForwardRef(withForwardedRef(Button))
    block={false}
    disabled={false}
    grouped={true}
    key=".1"
    label="Two"
    priority="outline"
    size="medium"
  />
</div>
`;

exports[`rendering renders correctly size 1`] = `
<div
  className="root"
  role="group"
>
  <ForwardRef(withForwardedRef(Button))
    block={false}
    disabled={false}
    grouped={true}
    key=".0"
    label="One"
    priority="default"
    size="large"
  />
  <ForwardRef(withForwardedRef(Button))
    block={false}
    disabled={false}
    grouped={true}
    key=".1"
    label="Two"
    priority="default"
    size="large"
  />
</div>
`;

exports[`rendering renders correctly with all props 1`] = `
<div
  className="root
        isRootBlock"
  role="group"
>
  <ForwardRef(withForwardedRef(Button))
    block={true}
    disabled={true}
    grouped={true}
    key=".0"
    label="One"
    priority="outline"
    size="small"
  />
  <ForwardRef(withForwardedRef(Button))
    block={true}
    disabled={true}
    grouped={true}
    key=".1"
    label="Two"
    priority="outline"
    size="small"
  />
</div>
`;
